### ISSUE TITLE
calcite-dropdown: Add additional placement options for left/right

### ISSUE THREAD

**Issue by ellenupp** [Esri/calcite-design-system#4124]
**monday.com sync:** #18048508861

### Description

Currently, the only [documented](https://developers.arcgis.com/calcite-design-system/components/dropdown/#component-api-properties-placement) placements for the dropdown are variations of top/bottom. We have a scenario where it would be ideal for the menu to open to the left or right instead. Setting any valid popper placement currently appears to work ([example](https://codepen.io/ellupp/pen/XWzzOwP?editors=1000)), but it would be nice if these were more officially accepted and documented as options. Didn't necessarily want to use these options as is if there's a chance they would be deprecated in the future since they aren't documented. 

### Acceptance Criteria

`left-*` and `right-*` are valid placement options for calcite-dropdown

**from Calcite team:**

* update all dropdown-related `placement: MenuPlacement` properties to use [`LogicalPlacement`](https://github.com/Esri/calcite-design-system/blob/main/packages/calcite-components/src/utils/floating-ui.ts#L212) instead

### Which Component

calcite-dropdown

### Example Use Case

![image](https://user-images.githubusercontent.com/14304581/154370826-23ef1cfa-570f-42a7-920d-8437c8d40e4c.png)

**Comment by driskull** [comment]
hey @ellenupp I'm not sure we want to support those. 

I think for these kind of menus like dropdown, combobox, etc. We only want to have them open above or below consistent with how native select menus appear.

@jcfranco @benelan Maybe we can further restrict them from even working if we are concerned with users setting them.

**Comment by ellenupp** [comment]
I guess my use case is not a typical use case for a dropdown, where it'd be in a form or such. These are floating menus that appear around each widget on the dashboard. That means their position will be highly variable (depends on dashboard layout & also browser window size) and there may or may not always be space for the menu to open above or below the trigger, in which case it'll need to open next to it to avoid being cut off. It'd just be really nice if the dropdown supported those placements because it otherwise works really well for these menus.

**Comment by vijayendranj** [comment]
@driskull Is there another UI calcite proposing in those scenarios?. simply restricting them from working is not a good user experience. Users use browsers in several ways  for ex. if the layout goes from landscape to portrait the height may be considerably less.

We'd be happy give a demo of how we are using this in Dashboard.

**Comment by ellenupp** [comment]
Here is a video showing the usage, including an example where left placement is preferable over top or bottom.
![2022-03-15_11-31-42 (3)](https://user-images.githubusercontent.com/14304581/158448913-0f44d745-d0b4-4869-a307-1a28c37d4c62.gif)

**Comment by jcfranco** [comment]
> Dropdown can be invoked from any trigger - where having it on the side as shown above makes sense.

@macandcheese @ashetland @SkyeSeitz Can we confirm if this is a supported use case from the UI/UX perspective? If so, we should widen the placement options.

According to our doc, [any element can be slotted as a trigger](https://developers.arcgis.com/calcite-design-system/components/dropdown/#component-api-slots-dropdown-trigger), so it's not clear what our dropdown trigger/placement guidance is.

**Comment by ashetland** [comment]
I agree this make a lot of sense. Since anything can be a trigger, the dropdown should be positionable to where it best fits the use case. Top / bottom / left / right + start / end would all be acceptable depending on the layout.

**Comment by driskull** [comment]
Yeah I think if we allow it on dropdown because its not a form element then that's fine. We should run it by the a11y team though just to make sure. We can ask them if a role of "menu" is allowed to open on the sides.

**Comment by jcfranco** [comment]
@geospatialem Can you help us get an answer for the above? I would assume that as long as the proper roles, ARIA attributes are used, and keyboard usage is maintained, the position of the menu won't be relevant, but I could be wrong. 

* https://www.w3.org/WAI/GL/wiki/Using_ARIA_menus
* https://www.w3.org/WAI/ARIA/apg/patterns/menu/
* https://www.w3.org/WAI/ARIA/apg/example-index/menubar/menubar-navigation
* https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/menu_role
* https://www.digitala11y.com/menu-role/

**Comment by geospatialem** [comment]
Somehow missed the notification on this, apologies for the late reply.

> Can you help us get an answer for the above? I would assume that as long as the proper roles, ARIA attributes are used, and keyboard usage is maintained, the position of the menu won't be relevant, but I could be wrong.

@jcfranco Yes, that is accurate. The placement of the component wouldn't impact assistive technologies or audiences, as long as the proper roles and the user experience is the expected pattern.

**Comment by macandcheese** [comment]
Since this already works (https://codepen.io/mac_and_cheese/pen/GRYVRBX?editors=1000), to get it documented I think we just need to use the type that has all the available options, not the limited set here: https://github.com/Esri/calcite-components/blob/4150b57358bcc180e298c9eaff48fbd2fea1c4a7/src/utils/floating-ui.ts#L170

**Comment by jcfranco** [comment]
@macandcheese @driskull WDYT about typing `placement` props with a suggested and fallback types? Something like the following:

```tsx
@Prop({ reflect: true }) placement: MenuPlacement | LogicalPlacement = defaultMenuPlacement;
```

We could get fancy with naming or excluded types if it makes things clearer in the doc and code too.

**Comment by driskull** [comment]
If we're not going to restrict "menus" to MenuPlacement then we can just get rid of that type and allow anything.

**Comment by jcfranco** [comment]
> If we're not going to restrict "menus" to MenuPlacement then we can just get rid of that type and allow anything.

Synced up w/ @driskull. We'll proceed with the above and update `placement`'s doc with recommended options for menus where applicable.

For context, adding [separate types](https://github.com/Esri/calcite-design-system/issues/4124#issuecomment-1843989236) isn’t that helpful in the code since IntelliSense flattens all types anyway, and the doc site would serve as the main reference for what these types allow if they were added.

**Comment by RSantosGIS** [comment]
Hey all, I was reviewing our todo items in our code base and wanted to chime in that we still have a standing issue tied to this.  We are currently using a type override for cases where we need the "left" or "right" options, which we use even though they aren't officially supported.  Code snippet below.  We use this for CalciteDropdown

![Image](https://github.com/user-attachments/assets/c28aa319-e64e-4494-a09c-4d668f001c4b)

**Comment by isaacbraun** [comment]
During verification, I see that PR #14132 adjusted the typings for `dropdown`, `split-button`, and `sort-handle`, but have three questions regarding the acceptance criteria:
- The `MenuPlacement` type is still used on three components (`alert`, `autocomplete`, and `input-date-picker`). Is keeping `MenuPlacement` on these components intentional?
- If the above is not intentional, do we still intend to fully remove the `MenuPlacement` type?
- The criteria says that the `MenuPlacement` placement values should be added to the JSDoc as suggestions, but I do not see that done for the components adjusted in #14132. Again, was this done intentionally or should they be updated?

TIA! cc @jcfranco @driskull

**Comment by driskull** [comment]
@isaacbraun I updated all the components that use dropdown. I didn't want to touch components outside of the dropdown scope. If we want to change the others it should probably be a new issue.

**Comment by jcfranco** [comment]
> - The `MenuPlacement` type is still used on three components (`alert`, `autocomplete`, and `input-date-picker`). Is keeping `MenuPlacement` on these components intentional?

I think `alert` needs a small update to alias the placements it uses. Although the values are the same, there is no menu context there. For the others, I think it is fine to keep, since they have a prescribed input and listbox relationship that still aligns with menu placement. `dropdown` feels like the special case, methinks, since its trigger can be anything and placement therefore needs to be more flexible.

> - If the above is not intentional, do we still intend to fully remove the `MenuPlacement` type?

Based on the above, I don't think so.

> - The criteria says that the `MenuPlacement` placement values should be added to the JSDoc as suggestions, but I do not see that done for the components adjusted in #14132. Again, was this done intentionally or should they be updated?

Not sure we need to anymore based on recent updates to dropdown (i.e., it supporting a wider set of configurations).

**Comment by isaacbraun** [comment]
Awesome, thank you Franco! 🤜💥🤛 Removed those two requirements from the acceptance criteria and opened #14304 to handle the Alert type alias.

With that said, this issue is verified on `5.1.0-next.36` via https://codepen.io/isaacbraun/pen/XJjwgGy.
- Dropdown and Split Button display the `LogicalPlacement` type for `placement`.
- All options are displayed as available on the component sample configuration, and they work via testing in the above CodePen. 

<img width="1012" height="84" alt="Image" src="https://github.com/user-attachments/assets/5513832a-bbe9-4404-a489-920837fed320" />

<img width="362" height="531" alt="Image" src="https://github.com/user-attachments/assets/92184acb-e248-413a-a778-f8eae24b7543" />
